physics/clm_lake.f90

```diff
@@ -0,0 +1,5607 @@
+             t_snow(c)  = t_snow(c) + t_soisno(c,j)
+             snowice(c) = snowice(c) + h2osoi_ice(c,j)
+             snowliq(c) = snowliq(c) + h2osoi_liq(c,j)
+          end if
+       end do
+    end do
+
+    ! Determine ending water balance and volumetric soil water
+
+    !dir$ concurrent
+    !cdir nodep
+    do fc = 1, num_shlakec
+       
+       c = filter_shlakec(fc)
+       if (snl(c) < 0) t_snow(c) = t_snow(c)/abs(snl(c))
+       endwb(c) = h2osno(c)
+    end do
+
+    do j = 1, nlevsoil
+      !dir$ concurrent
+      !cdir nodep
+       do fc = 1, num_shlakec
+          c = filter_shlakec(fc)
+          endwb(c) = endwb(c) + h2osoi_ice(c,j) + h2osoi_liq(c,j)
+          h2osoi_vol(c,j) = h2osoi_liq(c,j)/(dz(c,j)*denh2o) + h2osoi_ice(c,j)/(dz(c,j)*denice)
+       end do
+    end do
+
+    check_add_snow_water: if(LAKEDEBUG) then
+      allocate(snow_water(lbc:ubc))
+      ! Check to make sure snow water adds up correctly.
+      do j = -nlevsnow+1,0
+        !dir$ concurrent
+        !cdir nodep
+        do fc = 1, num_shlakec
+          c = filter_shlakec(fc)
+
+          jtop = snl(c)+1
+          if(j == jtop) snow_water(c) = 0._kind_lake
+          if(j >= jtop) then
+            snow_water(c) = snow_water(c) + h2osoi_ice(c,j) + h2osoi_liq(c,j)
+            if(j == 0 .and. abs(snow_water(c)-h2osno(c))>1.e-7_kind_lake) then
+              write(message,*)'h2osno does not equal sum of snow layers in ShalLakeHydrology:', &
+                   'column, h2osno, sum of snow layers =', c, h2osno(c), snow_water(c)
+              ! errmsg=trim(message)
+              ! errflg=1
+              write(0,'(A)') trim(message)
+            end if
+          end if
+        end do
+      end do
+      deallocate(snow_water)
+    end if check_add_snow_water
+
+    !!!!!!!!!!!!!
+    ! Do history variables and set special landunit runoff (adapted from end of HydrologyLake)
+    !dir$ concurrent
+    !cdir nodep
+    do fp = 1,num_shlakep
+       p = filter_shlakep(fp)
+       c = pcolumn(p)
+       g = pgridcell(p)
+
+       qflx_infl(c)      = 0._kind_lake
+       qflx_surf(c)      = 0._kind_lake
+       qflx_drain(c)     = 0._kind_lake
+       rootr_column(c,:) = spval
+       soilalpha(c)      = spval
+       zwt(c)            = spval
+       fcov(c)           = spval
+       qcharge(c)        = spval
+       !       h2osoi_vol(c,:)   = spval
+
+       ! Insure water balance using qflx_qrgwl
+       qflx_qrgwl(c)     = forc_rain(g) + forc_snow(g) - qflx_evap_tot(p) - (endwb(c)-begwb(c))/dtime
+       if (LAKEDEBUG) then
+         print *,'c, rain, snow, evap, endwb, begwb, qflx_qrgwl:', &
+              c, forc_rain(g), forc_snow(g), qflx_evap_tot(p), endwb(c), begwb(c), qflx_qrgwl(c)
+       endif
+
+       ! The pft average must be done here for output to history tape
+       qflx_evap_tot_col(c) = qflx_evap_tot(p)
+    end do
+    
+  end subroutine ShalLakeHydrology
+
+  subroutine QSat (T, p, es, esdT, qs, qsdT)
```

Comment:
@grantfirl note to self: another independent saturation algorithm

Reply:
This one is run at double precision (kind_lake=kind_dbl_prec)